vaadin-rich-text-editor-flow-parent/vaadin-rich-text-editor-flow/src/main/java/com/vaadin/flow/component/richtexteditor/RichTextEditor.java

```diff
@@ -955,40 +848,111 @@ public boolean isRequiredIndicatorVisible() {
             return RichTextEditor.this.isRequiredIndicatorVisible();
         }
 
-        private class HtmlValue implements Serializable {
-            private String value;
-            private boolean present;
+        @Override
+        public String getEmptyValue() {
+            return "";
+        }
+    }
 
-            private String getValue() {
-                if (!present) {
-                    this.value = generateHtmlValue();
-                    this.present = true;
-                }
-                return value;
+    private class AsDelta
+            implements HasValue<ValueChangeEvent<String>, String> {
+
+        private static class DeltaValueChangeEvent
+                extends ComponentEvent<RichTextEditor>
+                implements ValueChangeEvent<String> {
+
+            private final HasValue<?, String> hasValue;
+            private final String oldValue;
+
+            public DeltaValueChangeEvent(RichTextEditor source,
+                    HasValue<?, String> hasValue, String oldValue,
+                    boolean fromClient) {
+                super(source, fromClient);
+                this.hasValue = hasValue;
+                this.oldValue = oldValue;
+            }
+
+            @Override
+            public HasValue<?, String> getHasValue() {
+                return hasValue;
             }
 
-            private void setValue(String value) {
-                this.value = value;
-                this.present = true;
+            @Override
+            public boolean isFromClient() {
+                return super.isFromClient();
             }
 
-            private void clear() {
-                this.value = null;
-                this.present = false;
+            @Override
+            public String getOldValue() {
+                return oldValue;
             }
 
-            private String generateHtmlValue() {
-                if (RichTextEditor.this.isEmpty()) {
-                    return null;
+            @Override
+            public String getValue() {
+                return hasValue.getValue();
+            }
+        }
+
+        private boolean isHtmlValueSync;
+        private Registration deltaValueSyncRegistration;
+        private String oldValue = "";
+
+        public AsDelta() {
+            // Initialize empty value
+            RichTextEditor.this.getElement().setProperty("value", "");
+
+            // Fire delta value change event when HTML value changes
+            RichTextEditor.this.addValueChangeListener(event -> {
```

Comment:
![br](https://user-images.githubusercontent.com/357820/212734307-26ef7489-fba8-4b6d-b94f-efd8844272ba.png) 

> Now, let's have some fun!


This is where the fun starts! The previous implementation had a "sync" so that when you set the HTML value on the server, it would sync back the Delta value, and trigger a value change event for the Delta value.

This is now the equivalent: When the HTML value changes on the server, then we register a temporary property change listener for `value`, and wait for the client to synchronize it back. Afterwards we dispatch a value change event from `AsDelta`, as coming from the server. That means the delta value change event will only be fired after the roundtrip, but that was also the case previously - and there is no other way to do it, as the HTML <-> Delta conversion happens on the client.

If the HTML value has been changed on the client, we can immediately dispatch the delta value change event, as the `value` property has been synchronized in the same roundtrip (see the `getDeltaValue` method that is annotated with `@Synchronize`).

For both cases, value changes from client or from server, we delegate from the component's value changed event to an `AsDelta` specific `DeltaValueChangedEvent`. This is necessary, so that we can properly keep track of the `oldValue` that is necessary for future delta value change events. Updating the `oldValue` after firing the delta value change event guarantees that all change listeners for the delta value have been called, and it is safe to update it.